content/blog/2020-12-18-db-compiler/index.md

```diff
@@ -0,0 +1,293 @@
+
+I also verify correctness by checking that the query results are identical in Postgres.
+
+### Setup
+
+I'm running this on a Macbook Pro, Late 2013 with an i7-4750 HQ and 8GB of RAM. This is running Ubuntu 20.04 under the Windows Linux Subystem.
+
+For both systems, I took total running times on the client. THis means it includes time to print the result rows were printed but I redirected stdout to /dev/null. While this does include time to execute query planning for Postgres, Postgres reports that it takes less than a millisecond for these on all queries so I've omitted values beyond the decimal place to account for the loss of accuracy.
+
+### Results
+
+None of the queries had a correctness error.
+
+I ran 4 trials for each query and report the 95% (t-statistic) confidence intervals for the runtime. I report two columns for my system, one with the compile times omitted (so just the time spent actually processing the query) and also the total time for compilation and execution.
+
+| Query | Postgres Interval (ms) | KushDB Execution Only Interval (ms) | KushDB Total Interval (ms) |
+| ----  | -------------| -----------|------------|
+| Q1    | 2027 - 2332  | 192 - 466 | 1061 - 1611 |
+| Q3    | 818  - 1253  | 562 - 617 | 1771 - 2052 |
+| Q6    | 434   - 684  | 63 - 68   | 762 - 1502  |
+| Q11   | 118  - 186   | 28 - 118  | 1186 - 1554 |
+| Q12   | 506  - 696   | 105 - 114 | 922 - 1262  |
+
+
+![phase1](runtime.jpg)
+
+![phase1](total-runtime.jpg)
+
+For all queries generated, the compilation version (once compiled) greatly outperforms Postgres or matches it in the case of Q11. I suspect that the noise of measurement outweights any improvement in runtime given how fast both systems execute the query. However, once we take into account the compilation time, a different result emerges.
+
+### Compilation Time
+
+![phase1](compile.jpg)
+
+At such low table sizes, the time spent compiling actually dominates. Note that the time spent compiling is fixed so as the database size grows larger, the yellow portion of the graph will increase and the grey portion will stay constant. This means the ratio of time spent compiling to time spent executing will decrease on larger databases.
+
+However, this is still unacceptable on smaller data sets where Postgres outperforms the compilation version. In future work, I discuss some ways that this can be fixed.
+
+## Personal Experiences
+
+The largest barrier I had to overcome in working on this project was understanding all the pieces of the puzzle. I needed to understand each portion of the system enough to implement it.
+
+## Future Work
+
+While this generates extremely high performant code, the time spent compiling is extremely large. The following strategies are useful in reducing this:
+- Directly generate LLVM. This removes any overhead for file I/O to output the C++ and parsing, type checking and converting that C++ to LLVM.
```

Comment:
FWIW, the conventional wisdom is that LLVM might not be a great fit for most JITs, at least before the highest tier, because it is built to prioritize optimal code over compilation time. So many JITs use simpler code generators such as [DynASM](https://luajit.org/dynasm.html), [B3](https://webkit.org/docs/b3/), or [Cranelift](https://cranelift.readthedocs.io/en/latest/). The same probably applies to query compilers—after all, it is basically a JIT.

Reply:
Gave a little more context here.